src/ensembl/src/content/app/entity-viewer/gene-view/components/protein-domain-image/ProteinDomainImage.tsx

```diff
@@ -0,0 +1,170 @@
+import React from 'react';
+import classNames from 'classnames';
+import { scaleLinear, ScaleLinear } from 'd3';
+
+import { getFeatureCoordinates } from 'src/content/app/entity-viewer/shared/helpers/entity-helpers';
+import { Transcript } from 'src/content/app/entity-viewer/types/transcript';
+import { ProteinDomainsResources } from 'src/content/app/entity-viewer/types/product';
+
+import styles from './ProteinDomainImage.scss';
+
+const BLOCK_HEIGHT = 18;
+const TRACK_HEIGHT = 24;
+
+export type ProteinDomainImageProps = {
+  transcript: Transcript;
+  width: number; // available width for drawing, in pixels
+  classNames?: {
+    track?: string;
+    domain?: string;
+  };
+};
+
+type ProteinDomainImageData = {
+  [resource_type: string]: {
+    [resource_description: string]: {
+      start: number;
+      end: number;
+    }[];
+  };
+};
+
+export const getDomainsByResourceGroups = (
+  proteinDomainsResources: ProteinDomainsResources
+) => {
+  const proteinDomains: ProteinDomainImageData = {};
+
+  Object.values(proteinDomainsResources).map((resource) => {
```

Comment:
Am I correct in understanding that everything inside this function are side effects to populate the `proteinDomains` object? If so, please use forEach instead of map. forEach is for side effects; map is for returning the transformed array of values.

Reply:
What is also strange about this line is that it's calling Object.values, and then not really using the values themselves. The only way that these values are used in the body of the function is to access their `name` property, which, in the way it's subsequently used, seems to be the same as keys on the proteinDomainsResources object.

Wouldn't it be less confusing to call Object.keys on `proteinDomainsResources`, and to use the keys themselves?

Reply:
Changed it to use forEach and Object.keys. I think I was using the values initially but changed the logic later on.